util/libav: do not bail on MEL-only RPUs w/EL

intelfx opened the pull request:
We can still map DV metadata that indicates enhancement layer usage (!disable_residual_flag, Profile 7 and 4) for MEL-only bitstreams, which is indicated by trivial NLQ parameters.

intelfx committed:
util/libav: do not bail on MEL-only RPUs w/EL

We can still map DV metadata that indicates enhancement layer usage
(!disable_residual_flag, Profile 7 and 4) for MEL-only bitstreams,
which is indicated by trivial NLQ parameters.

kasper93 commented:
`pl_map_avframe_ex()` should also be fixed in similar fashion to not use `header->disable_residual_flag`. And for parity with libdovi code path `max_pq` and `avg_pq` should be mapped also when dovi meta is not used.

``` diff
diff --git a/src/include/libplacebo/utils/libav_internal.h b/src/include/libplacebo/utils/libav_internal.h
index a754e98b..928d37f6 100644
--- a/src/include/libplacebo/utils/libav_internal.h
+++ b/src/include/libplacebo/utils/libav_internal.h
@@ -925,14 +925,14 @@ PL_LIBAV_API void pl_map_avdovi_metadata(struct pl_color_space *color,
             pl_hdr_rescale(PL_HDR_PQ, PL_HDR_NITS, dovi_color->source_min_pq / 4095.0f);
         color->hdr.max_luma =
             pl_hdr_rescale(PL_HDR_PQ, PL_HDR_NITS, dovi_color->source_max_pq / 4095.0f);
+    }

 #if LIBAVUTIL_VERSION_INT >= AV_VERSION_INT(59, 12, 100)
-        if ((dovi_ext = av_dovi_find_level(metadata, 1))) {
-            color->hdr.max_pq_y = dovi_ext->l1.max_pq / 4095.0f;
-            color->hdr.avg_pq_y = dovi_ext->l1.avg_pq / 4095.0f;
-        }
-#endif
+    if ((dovi_ext = av_dovi_find_level(metadata, 1))) {
+        color->hdr.max_pq_y = dovi_ext->l1.max_pq / 4095.0f;
+        color->hdr.avg_pq_y = dovi_ext->l1.avg_pq / 4095.0f;
     }
+#endif
 }

 PL_LIBAV_API void pl_frame_map_avdovi_metadata(struct pl_frame *out_frame,
```

which makes it awkward with `dovi` allocation part, but since the API is changed already it can be done while at it. See how `pl_map_avframe_ex()` maps RPU info always, regardless of `header->disable_residual_flag`.

kasper93 commented on src/include/libplacebo/utils/libav.h:
```diff
@@ -131,12 +131,18 @@ PL_DEPRECATED_IN(v7.343) PL_LIBAV_API void pl_frame_map_avdovi_metadata(
+// Helper function to check if Dolby Vision metadata can be mapped.
+// Returns true if the `AVDOVIRpuDataHeader` `disable_residual_flag` field is
+// not zero or if the NLQ parameters are trivial.
+PL_LIBAV_API bool pl_can_map_avdovi_metadata(const AVDOVIMetadata *metadata);
```
`pl_map_avdovi_metadata` could map more metadata, but without `pl_dovi_metadata` use. This could be `pl_is_dovi_metadata_needed` or something and `pl_map_avdovi_metadata` could take NULL if the former function returns false. This would allow to still make RPU values.

kasper93 replied:
`pl_map_avdovi_metadata(&color, &repr, NULL, &metadata);` I'm talking about output parameter, when there we decide not to use `pl_dovi_metadata`, but still populate `max_pq` and `avg_pq`. This is how `pl_map_avframe_ex()` works.

intelfx committed:
util/libav: lift supported checks out of `pl_map_avdovi_metadata()`

(But not `pl_frame_map_avdovi_metadata()`.)

Let caller decide whether they want to map the metadata. This way,
the caller can decide whether they want to use the luma and PQ values
even despite not having the NLQ curve and the EL actually processed.

intelfx committed:
util/libav: add `pl_avframe_params.map_dovi_force`

intelfx committed:
utils/libav: bump API version due to *avdovi_metadata changes

kasper93 reviewed (approved):
> I have split the suggested changes into their own commits just for visibility. I'll squash them back eventually. I'm not sure how to pass pl_options to `pl_map_avframe_ex()`; did I use the right options here or did you mean anything else?

I think using params is fine.

> I did bump the API version, but I'm not sure how I should handle the API semantics change. Do I declare a new function with new semantics and deprecate the old one, or should I leave it as it is now?

Which functions would you deprecate? 

We could deprecate libdovi usage.

> I also did not change the semantics of the already-deprecated `pl_frame_map_avdovi_metadata()`.

That's fine.

I think the patch is fine, it will change mapping for users that directly map dovi metadata using those API, but we have api version bump and generally preserving some backward compatibility with not mapping this metadata, would require duplicating APIs. I don't think it's worth it.

/cc @haasn

intelfx commented:
Updated the mpv counterpart in mpv-player/mpv#14692.

intelfx commented:
@kasper93, @haasn, is there something that needs to happen before this can be merged, or was it decided that this is a bad idea/worthless fix for some general reason?

kasper93 commented:
> @kasper93, @haasn, is there something that needs to happen before this can be merged, or was it decided that this is a bad idea/worthless fix for some general reason?

I've pulled your changes into https://code.videolan.org/videolan/libplacebo/-/merge_requests/832, because we have CI there, and patches are merged through gitlab.

On my end, the change looks ok, but I'd like @haasn to comment on this topic, last time he had suggestions to API, maybe this needs to be pursued first. For me current version is fine.

kasper93 commented:
Merged as d5ef02c7dfba5bb2acee0aaabe5861b7b3abc073. Thanks! Sorry for the delay.
